midend/lib/Utils/DIPUtils.cpp

```diff
@@ -410,6 +411,185 @@ Value customTanVal(OpBuilder &builder, Location loc, Value angleVal) {
   return builder.create<arith::DivFOp>(loc, sinVal, cosVal);
 }
 
+SmallVector<Value, 6> getRotationMatrix(OpBuilder& builder, Location loc, Value centerX, Value centerY, Value angle, Value scale) {
+  Value alpha0 = builder.create<math::CosOp>(loc, angle);
+  Value alpha = builder.create<arith::MulFOp>(loc, alpha0, scale);
+  Value beta0 = builder.create<math::SinOp>(loc, angle);
+  Value beta = builder.create<arith::MulFOp>(loc, beta0, scale);
+  Value oneMinusAlpha = builder.create<arith::SubFOp>(loc, builder.create<arith::ConstantOp>(loc, builder.getF32FloatAttr((float)1.)), alpha);
+  SmallVector<Value, 6> mat;
+  Value m20 = builder.create<arith::MulFOp>(loc, oneMinusAlpha, centerX);
+  Value m21 = builder.create<arith::MulFOp>(loc, beta, centerY);
+  Value m50 = builder.create<arith::MulFOp>(loc, beta, centerX);
+  Value m51 = builder.create<arith::MulFOp>(loc, oneMinusAlpha, centerY);
+  Value m0 = alpha;
+  Value m1 = beta;
+  Value m2 = builder.create<arith::SubFOp>(loc, m20, m21);
+  Value m3 = builder.create<arith::NegFOp>(loc, beta);
+  Value m4 = alpha;
+  Value m5 = builder.create<arith::AddFOp>(loc, m50, m51);
+
+  return SmallVector<Value, 6>{m0, m1, m2, m3, m4, m5};
+}
+
+inline void inverseAffineMatrix(OpBuilder &builder, Location loc,
+                                SmallVector<Value, 6> &affineMatrix) {
+  Value c0F32 = builder.create<arith::ConstantOp>(
+      loc, builder.getF32FloatAttr((float).0));
+  Value m0pm4 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[0], affineMatrix[4]);
+  Value m1pm3 =
+      builder.create<arith::MulFOp>(loc, affineMatrix[1], affineMatrix[3]);
+  Value D = builder.create<arith::SubFOp>(loc, m0pm4, m1pm3);
+  // TODO: can a float number compare with 0.0 directly?
+  Value dEq0 =
+      builder.create<arith::CmpFOp>(loc, arith::CmpFPredicate::OEQ, D, c0F32);
+  auto scfRes = builder.create<scf::IfOp>(
+      loc, TypeRange{builder.getF32Type()}, dEq0,
+      [&](OpBuilder &thenBuilder, Location thenLoc) {
+        thenBuilder.create<scf::YieldOp>(thenLoc, ValueRange{c0F32});
+      },
+      [&](OpBuilder &elseBuilder, Location elseLoc) {
+        Value c1F32 = elseBuilder.create<arith::ConstantOp>(
+            elseLoc, builder.getF32FloatAttr((float)1.));
+        Value res = elseBuilder.create<arith::DivFOp>(elseLoc, c1F32, D);
+        elseBuilder.create<scf::YieldOp>(elseLoc, ValueRange{res});
+      });
+  D = scfRes.getResult(0);
+  Value negD = builder.create<arith::NegFOp>(loc, D);
+  Value A11 = builder.create<arith::MulFOp>(loc, affineMatrix[4], D);
+  Value A22 = builder.create<arith::MulFOp>(loc, affineMatrix[0], D);
+  affineMatrix[0] = A11;
```

Comment:
Why not directly assign the operation result itself, instead of introducing `A11`?
Same argument for `A22` as well.